InvenTree/machine/models.py

```diff
@@ -0,0 +1,150 @@
+from typing import Any
+
+from django.contrib import admin
+from django.db import models
+from django.utils.html import format_html_join
+from django.utils.safestring import mark_safe
+from django.utils.translation import gettext_lazy as _
+
+import common.models
+from machine.registry import registry
+
+
+class Machine(models.Model):
+    """A Machine objects represents a physical machine."""
+
+    name = models.CharField(
+        unique=True,
+        max_length=255,
+        verbose_name=_("Name"),
+        help_text=_("Name of machine")
+    )
+
+    machine_type_key = models.CharField(
+        max_length=255,
+        verbose_name=_("Machine Type"),
+        help_text=_("Type of machine"),
+    )
+
+    driver_key = models.CharField(
+        max_length=255,
+        verbose_name=_("Driver"),
+        help_text=_("Driver used for the machine")
+    )
+
+    active = models.BooleanField(
+        default=True,
+        verbose_name=_("Active"),
+        help_text=_("Machines can be disabled")
+    )
+
+    def __str__(self) -> str:
+        """String representation of a machine."""
+        return f"{self.name}"
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Override to set origignal state of the machine instance."""
+        super().__init__(*args, **kwargs)
+
+        self.errors = []
+        self.initialized = False
```

Comment:
What is the best way to store such state as these model instances are recreated everytime I use them and therefor this state is not preserved.

Reply:
I am not sure if the state should be saved to the database. What if the device changes while the instance is down? I think the state should be processed on initialization of the machine and the be dynamic.

Reply:
No, this is no model field, doesn't this work with Django to have some fields that are stored only during runtime?

Reply:
@wolflu05 the idea would be to do several steps on initialisation:
- load all database objects for machines
- issue 'machine_registry_init_start' event
- iterate over the objects
  - init machine from machine type with db object
  - here is the initial state set on the machine object
  - run a machine setup method to give it the option to discover "real" state
  - then save the machine object (not db ref) to the registry
- issue 'machine_registry_init_end' event